src/GrafanaDashboardDeploy/src/DeployTool.cs

```diff
@@ -0,0 +1,152 @@
+using Microsoft.Azure.KeyVault;
+using Microsoft.Azure.Services.AppAuthentication;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace DotNet.Grafana
+{
+    public static class DeployTool
+    {
+        public static async Task PostExportPack(GrafanaClient grafanaClient, string inputFilePath)
+        {
+            var folderIdMap = new Dictionary<string, int>();
+
+            using var sr = new StreamReader(inputFilePath);
+            using var jr = new JsonTextReader(sr);
+            var data = await JObject.LoadAsync(jr).ConfigureAwait(false);
+
+            foreach (var folder in data["folders"])
+            {
+                string folderUid = folder["uid"].Value<string>();
+                string folderTitle = folder["title"].Value<string>();
+
+                var result = await grafanaClient.CreateFolderAsync(folderUid, folderTitle).ConfigureAwait(false);
+                int folderId = result["id"].Value<int>();
+
+                folderIdMap.Add(folderUid, folderId);
+            }
+
+            foreach (var datasource in data["datasources"])
+            {
+                foreach (var secretData in datasource["secureJsonData"].Children<JProperty>())
+                {
+                    string secretExpression = secretData.Value.ToString();
+                    string secretName;
+                    string secretValue;
+
+                    if (!TryGetSecretName(secretExpression, out secretName))
+                    {
+                        continue;
+                    }
+
+                    secretValue = await GetSecretAsync(secretName).ConfigureAwait(false);
+                    secretData.Value = secretValue;
+                }
+
+                await grafanaClient.CreateDatasourceAsync(datasource.ToObject<JObject>()).ConfigureAwait(false);
+            }
+
+            foreach (var dashboard in data["dashboards"])
+            {
+                string dashboardFolderUid = dashboard["meta"]["folderUid"].Value<string>();
+
+                int folderId = folderIdMap[dashboardFolderUid];
+
+                await grafanaClient.CreateDashboardAsync(dashboard["dashboard"].ToObject<JObject>(), folderId).ConfigureAwait(false);
```

Comment:
This line is a bit long... and it's odd to "ToObject<JObject>" these things should already _be_ JObjects... deserializing it into what it already is is... unexpected.

Reply:
The type returned from indexing that way is JToken. I landed with JObject as a better base object to pass around the library, which requires some conversions.

Reply:
JObject extends JToken, you can just cast it. You don't need to re-serialize it.